### ISSUE TITLE
Image level bounding boxes in YOLOWorld

### ISSUE THREAD

**Issue by toobatehreem** [ultralytics/ultralytics#17976]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How can I get image level predicted and ground truth bounding boxes during the forward pass. Maybe within the v8Detection loss function. As the current predictions and gt bboxes seem to be pixel level and after mosaic transformation.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @toobatehreem, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it.

If this is a custom feature ❓ Question (such as retrieving image-level predicted and ground truth bounding boxes) or involves modifying the codebase, please provide as much context as possible, including the specific part of the code you're working with, as well as detailed examples or any relevant logs. Additionally, ensure you're familiar with how augmentations like mosaic impact data preprocessing and predictions by reviewing our [Training Tips](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best! For real-time discussions, check out [Discord](https://discord.com/invite/ultralytics) 🎧, engage with in-depth threads on [Discourse](https://community.ultralytics.com), or connect with others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Please ensure you are using the latest version of the `ultralytics` package to avoid issues. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed:

```bash
pip install -U ultralytics
```

## Environments

For experimenting in verified environments with pre-installed dependencies, you can utilize:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. CI tests validate correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu every 24 hours and on every commit.

An Ultralytics engineer will review your question soon to provide further insights. Thanks again for reaching out and for using Ultralytics tools! 🚀

**Comment by glenn-jocher** [comment]
@toobatehreem to obtain image-level bounding boxes during training in YOLOWorld, you will need to access the predicted and ground truth bounding boxes before augmentation (like mosaic). You can customize the dataloader or modify the model's training loop to extract them at this stage. Alternatively, you can refer directly to the `v8Detection` loss function, inspecting the processed outputs, and backtrack to unapply augmentations. For guidance, review the official YOLO source code or raise specific implementation queries in [discussions](https://github.com/ultralytics/ultralytics/discussions).

**Comment by Y-T-G** [comment]
You would have to create a new key in the labels object that stores the original boxes by cloning the value in `mix_labels`
https://github.com/ultralytics/ultralytics/blob/461597e07cd457224a2fb179d719e4d235529c14/ultralytics/data/augment.py#L593

and make sure it's not deleted during augmentation, especially when it creates a new `Instances` object.

https://github.com/ultralytics/ultralytics/blob/461597e07cd457224a2fb179d719e4d235529c14/ultralytics/data/augment.py#L1246

**Comment by toobatehreem** [comment]
> @toobatehreem to obtain image-level bounding boxes during training in YOLOWorld, you will need to access the predicted and ground truth bounding boxes before augmentation (like mosaic). You can customize the dataloader or modify the model's training loop to extract them at this stage. Alternatively, you can refer directly to the `v8Detection` loss function, inspecting the processed outputs, and backtrack to unapply augmentations. For guidance, review the official YOLO source code or raise specific implementation queries in [discussions](https://github.com/ultralytics/ultralytics/discussions).

I have two questions: 
1. What is the format of pred boxes after the forward pass inside the v8Detection loss?
2. Even if I pass the original gt boxes in the labels, I will still need to revert the preds to match them in the loss function, any idea on how to do that? @glenn-jocher @Y-T-G

**Comment by toobatehreem** [comment]
> @toobatehreem to obtain image-level bounding boxes during training in YOLOWorld, you will need to access the predicted and ground truth bounding boxes before augmentation (like mosaic). You can customize the dataloader or modify the model's training loop to extract them at this stage. Alternatively, you can refer directly to the `v8Detection` loss function, inspecting the processed outputs, and backtrack to unapply augmentations. For guidance, review the official YOLO source code or raise specific implementation queries in [discussions](https://github.com/ultralytics/ultralytics/discussions).

Also, can I use the boxes decoding code written for the inference in the WorldDetect class in head.py for the predictions I get during training? If not, what's the difference between both.
This the box decoding box during inference: # Inference path

```python
            shape = x[0].shape  # BCHW
            x_cat = torch.cat([xi.view(shape[0], self.nc + self.reg_max * 4, -1) for xi in x], 2)
            if self.dynamic or self.shape != shape:
                self.anchors, self.strides = (x.transpose(0, 1) for x in make_anchors(x, self.stride, 0.5))
                self.shape = shape

            if self.export and self.format in {"saved_model", "pb", "tflite", "edgetpu", "tfjs"}:  # avoid TF FlexSplitV ops
                box = x_cat[:, : self.reg_max * 4]
                cls = x_cat[:, self.reg_max * 4 :]
            else:
                box, cls = x_cat.split((self.reg_max * 4, self.nc), 1)

            if self.export and self.format in {"tflite", "edgetpu"}:
                # Precompute normalization factor to increase numerical stability
                # See https://github.com/ultralytics/ultralytics/issues/7371
                grid_h = shape[2]
                grid_w = shape[3]
                grid_size = torch.tensor([grid_w, grid_h, grid_w, grid_h], device=box.device).reshape(1, 4, 1)
                norm = self.strides / (self.stride[0] * grid_size)
                dbox = self.decode_bboxes(self.dfl(box) * norm, self.anchors.unsqueeze(0) * norm[:, :2])
            else:
                dbox = self.decode_bboxes(self.dfl(box), self.anchors.unsqueeze(0)) * self.strides

            y = torch.cat((dbox, cls.sigmoid()), 1)
            return y if self.export else (y, x)
```

**Comment by glenn-jocher** [comment]
Great question! Yes, you can utilize the decoding approach from the inference path in the `WorldDetect` class for visualizing the predictions during training since both methods ultimately aim to decode the distributional predictions into boxes. However, there are differences based on context and stage:

1. **Inference Decoding**: This focuses on generating final decoded boxes for visualization or downstream tasks (e.g., NMS is applied later, and boxes are scaled to image dimensions). The code uses class probabilities (`cls.sigmoid()`) and anchors adjusted to strides.
   
2. **Loss Decoding**: This operates within the training loop to calculate losses. It decodes boxes with respect to anchor points (via `bbox_decode`), which are then matched against ground truth (GT). The transformations ensure predictions and GT bboxes align for effective loss computation.

While the decoding steps share similarities, inference introduces additional post-processing steps (e.g., applying softmax/sigmoid directly and scaling). For training, stick to the `bbox_decode` in `v8Detection` loss for accuracy and synchronization with GT. To better understand both processes, review anchor generation (`make_anchors`) and decoding functions directly in the source code. Let me know if further clarification is needed!

**Comment by toobatehreem** [comment]
> Great question! Yes, you can utilize the decoding approach from the inference path in the `WorldDetect` class for visualizing the predictions during training since both methods ultimately aim to decode the distributional predictions into boxes. However, there are differences based on context and stage:
> 
> 1. **Inference Decoding**: This focuses on generating final decoded boxes for visualization or downstream tasks (e.g., NMS is applied later, and boxes are scaled to image dimensions). The code uses class probabilities (`cls.sigmoid()`) and anchors adjusted to strides.
> 2. **Loss Decoding**: This operates within the training loop to calculate losses. It decodes boxes with respect to anchor points (via `bbox_decode`), which are then matched against ground truth (GT). The transformations ensure predictions and GT bboxes align for effective loss computation.
> 
> While the decoding steps share similarities, inference introduces additional post-processing steps (e.g., applying softmax/sigmoid directly and scaling). For training, stick to the `bbox_decode` in `v8Detection` loss for accuracy and synchronization with GT. To better understand both processes, review anchor generation (`make_anchors`) and decoding functions directly in the source code. Let me know if further clarification is needed!

Thank you.

**Comment by Y-T-G** [comment]
The loss receives the augmented boxes. You can visualize the image by permuting the dims to H,WC and then denormalizing it by multiplying with 255 and then saving it.

**Comment by toobatehreem** [comment]
> The loss receives the augmented boxes. You can visualize the image by permuting the dims to H,W,C and then denormalizing it by multiplying with 255 and then saving it.

yes, I am able to visualize it inside the loss class. But now I need to take that predicted bounding box from the augmented image and visualize it on the original image (without augmentations). So I want to convert those bounding boxes to the original image level, and since mosaic combines 4 different images together, I have to de-augment it for each of the images in one mosaic sample. Any idea how to do it?

**Comment by glenn-jocher** [comment]
@toobatehreem to map the predicted bounding boxes back to the original image prior to augmentations like mosaic, you need to reverse the augmentation transformations applied during preprocessing. Specifically for mosaic:

1. Identify the placement of each original image within the mosaic (e.g., top-left, top-right, etc.).  
2. Adjust the bounding box coordinates by subtracting the offsets and scaling to the dimensions of the original image.  
3. Apply any unnormalization, if required (e.g., if bounding boxes were scaled to 0-1 range).  

You can refer to the mosaic implementation in `ultralytics/data/augment.py` to better understand how to reverse the specific augmentations applied. Let me know if additional clarification is needed!
